# An Analysis of Parking and Traffic Violations in New York City in 2022

# Table of Contents
1. [Introduction](#introduction)
2. [Skills Used](#skills-used)
3. [Objective](#objective)
4. [Data](#data)
5. [Procedure](#procedure)
6. [Loading the Data](#loading-the-data)
7. [Data Cleaning and Preparation](#data-cleaning-and-preparation)
    1. [Convert violation_time column to 24-hour format](#convert-violation_time-column-to-24-hr-format)
    2. [Extract the day of the week from the issue_date column](#extract-the-day-of-the-week-from-the-issue_date-column)
    3. [Create a Column for the Violation Address](#create-a-column-for-the-violation-address)
    4. [Convert Violation Addresses to Longitude and Latitude Coordinates](#convert-violation-addresses-to-longitude-and-latitude-coordinates)
8. [Explore the Data](#explore-the-data)
    1. [Descriptive Statistics](#descriptive-statistics)
    2. [The Top Violations](#the-top-violations)
    3. [Violations by Registration State](#violations-by-registration-state)
    4. [Time Series Analysis of Violations and Fines](#time-series-analysis-of-violations-and-fines)
    5. [Issuing Agency Analysis](#issuing-agency-analysis)
    6. [Vehicle Analysis](#vehicle-analysis)
9. [Dashboard Creation](#dashboard-creation)
10. [Conclusion](#conclusion)


# Introduction

New York City (NYC) is one of the most traffic-congested cities in the world. Consequently, NYC motorists frequently encounter parking and traffic violation issues. This project aims to explore parking and traffic violation data from NYC for the year 2022 to inform stakeholders about the extent of the problem and provide insights to help improve the situation.

This project is divided into two parts:
1. The first part, documented in this Jupyter Notebook, involves cleaning and exploring the data using SQL.
2. The second part involves using Tableau to visualize the data through charts and an interactive dashboard that allows users to interact with the data.

The Tableau visualizations and dashboard can be found by [clicking this link](https://public.tableau.com/views/ParkingandTrafficViolationsinNewYorkCity2022/NYCParkingandTrafficViolationsDashboard?:language=en-US&:sid=&:display_count=n&:origin=viz_share_link).

---

Key points to consider in this analysis:
- **Data Cleaning:** Ensuring the data is accurate and ready for analysis.
- **Exploratory Data Analysis (EDA):** Understanding the distribution, trends, and anomalies in the data.
- **Visualizations:** Creating informative and interactive visual representations of the data to highlight key insights.

By combining SQL for data processing and Tableau for data visualization, we aim to provide a comprehensive overview of parking and traffic violations in NYC, identify trends, and suggest possible areas for improvement.
   

# Skills Used
- SQL (MYSQL)
- Tableau Desktop
- Data Cleaning
- Data Visualization

In this project, I will use MySQL to query and analyze a large volume of parking and traffic violation data from NYC Open Data. I will then use Tableau to create a dashboard and visualize the insights obtained from the SQL queries on the data.

---

### Key Skills Highlighted:
- **SQL (MySQL):** For data querying, manipulation, and analysis. This includes writing complex queries to extract meaningful information from a large dataset.
- **Tableau:** For creating interactive and visually appealing charts and dashboards that help in interpreting and presenting the data insights effectively.

By leveraging these skills, the project aims to provide a comprehensive analysis of parking and traffic violations in NYC, uncovering trends and patterns that can inform decision-making and policy formulation.


# Objective

The main objective of this project is to explore NYC traffic violation data for the year 2022 to provide answers to the following questions:
* **Who were the main contributors to parking violations in NYC based on the state of vehicle registration?**
* **What was the distribution of parking violations across NYC? Which areas were the violation hotspots?**
* **How did parking violations vary throughout the year? Which months had the most violations?**
* **How did parking violations vary based on the time of day? At what time of day are you most likely to get ticketed?**
* **Which enforcement agency issued the most parking violations?**
* **What was the total amount of fines issued in 2022?**
* **What was the percentage of unpaid fines in 2022?**

---

### Analysis Insights:
- **Main Contributors:** Identify states from which the most violating vehicles are registered.
- **Geographical Distribution:** Determine hotspots and areas with the highest violation rates.
- **Temporal Trends:** Analyze monthly and daily patterns in parking violations.
- **Time of Day:** Assess the likelihood of receiving a ticket at different times.
- **Enforcement Agency Performance:** Evaluate which agencies are most active in issuing violations.
- **Financial Impact:** Calculate the total fines issued and the proportion that remains unpaid.

By addressing these questions, the project aims to provide a detailed understanding of parking violations in NYC, highlighting key trends and areas for potential policy improvement.





# Data

The data used in this project was obtained from NYC OpenData. The data was extracted for the year 2022 from two databases related to parking violations in NYC:

1. [Parking Violations Issued - Fiscal Year 2023](https://data.cityofnewyork.us/City-Government/Parking-Violations-Issued-Fiscal-Year-2023/869v-vr48/about_data)
2. [Open Parking and Camera Violations](https://data.cityofnewyork.us/City-Government/Open-Parking-and-Camera-Violations/nc67-uf89/about_data)

In both cases, the data was filtered to extract records for the year 2022. The data dictionaries for both datasets can be found at the provided links. 

In this project, most of the analysis is conducted using the Open Parking and Camera Violations dataset, which is stored in a table called `vio_22_time`. The dataset from the Parking Violations Issued - Fiscal Year 2023 is stored in a table called `parking_violations`.

---

### Data Overview:
- **Data Source:** NYC OpenData
- **Year of Analysis:** 2022
- **Primary Datasets:**
  - `Open Parking and Camera Violations` (Table: `vio_22_time`)
  - `Parking Violations Issued - Fiscal Year 2023` (Table: `parking_violations`)
  
By leveraging these datasets, we aim to conduct a thorough analysis of parking violations in NYC, providing valuable insights and trends from the year 2022.


# Procedure

The datasets, in the form of CSV files, were downloaded from NYC OpenData and then loaded into MySQL Workbench for analysis. In this Notebook, `mysql.connector` and the `pandas` library are used to connect to the database and to visualize the outcome of some of the queries in the form of DataFrames. The data were first cleaned and prepared using SQL queries. Exploratory Data Analysis (EDA) was then conducted to gain insights based on the questions identified earlier. 

Below, we outline the steps followed to load, clean, and prepare the data for analysis. The datasets were very similar in nature, and the steps documented below were followed for both datasets. The visualization of charts and a dashboard are presented separately in [at this link in Tableau Public](https://public.tableau.com/app/profile/shawn.best/viz/ParkingandTrafficViolationsinNewYorkCity2022/NYCParkingandTrafficViolationsDashboard).

### Key Steps in Data Preparation:
1. **Loading Data:** Importing CSV files into MySQL Workbench.
2. **Connecting to Database:** Using `mysql.connector` and `pandas` to establish a connection to the MySQL database.
3. **Data Cleaning:** Executing SQL queries to clean the data, including handling missing values, correcting data types, and removing duplicates.
4. **Data Preparation:** Preparing the data for analysis by creating necessary columns and tables.
5. **Exploratory Data Analysis (EDA):** Conducting an initial analysis to understand data distributions, identify trends, and uncover anomalies.
6. **Data Visualization:** Presenting the cleaned and analyzed data through interactive charts and dashboards in Tableau Public.

By following these steps, we ensure the data is properly prepared and analyzed, providing valuable insights into parking and traffic violations in NYC for the year 2022.


# Loading the Data

In [1]:
# Import libraries and load CSV files
import mysql.connector
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

def execute_query(query, user='', password='', host='', database='nyc_parking_vio'):
    
    # Establish a connection
    connection = mysql.connector.connect(
        host=host,
        user=user,
        password=password,
        database=database
    )
    
    # Execute the query and read the data into a pandas DataFrame
    df = pd.read_sql(query, connection)
    
    # Close the connection
    connection.close()
    
    return df



I encountered challenges while creating the table for the database. Although the data dictionary indicated that some columns were stored as numbers, I initially had to load some of the data as `VARCHAR(255)` and then convert the relevant columns to their respective data types. To manage this, a staging table (`parking_violations_staging`) was used to load the data. 

The `issue_date` column was then converted to a date format using `STR_TO_DATE(issue_date, '%m/%d/%Y')`. After the necessary conversions, the final table (`parking_violations`) was created, and the data was inserted into this table. Missing values in all columns were replaced with `NULL` values.

### Key Steps in Data Handling:
1. **Creating Staging Table:** Load initial data into a staging table (`parking_violations_staging`) with all columns as `VARCHAR(255)`.
2. **Data Type Conversion:** Convert relevant columns to their appropriate data types, such as converting `issue_date` to a date format using `STR_TO_DATE(issue_date, '%m/%d/%Y')`.
3. **Final Table Creation:** Create the final table (`parking_violations`) with the correct data types for each column.
4. **Data Insertion:** Insert cleaned and converted data from the staging table into the final table.
5. **Handling Missing Values:** Replace missing values in all columns with `NULL`.

By following these steps, I ensured the data was accurately loaded, transformed, and stored, enabling reliable analysis and visualization in subsequent steps.




Two data tables were created in MySQL Workbench for use in this analysis because the original datasets contained different aspects of NYC parking and traffic violations. The database is called `nyc_parking_vio`, and the tables used for analysis are `parking_violations` and `vio_22_time`.

### Key Details:
1. **Database Name:** `nyc_parking_vio`
2. **Tables for Analysis:**
   - `parking_violations`: Contains data related to parking violations.
   - `vio_22_time`: Contains data related to parking violations with time-specific information.

By organizing the data into these two tables, we can effectively analyze the various aspects of parking and traffic violations in NYC for the year 2022.
 

In [2]:
# Verify data transfer

query = """
SELECT 
    *
FROM
    parking_violations
LIMIT 10;
"""

df = execute_query(query)

# Display the DataFrame
df.head()



summons_number plate_id registration_state plate_type issue_date  \
0      1484697303  JER1863                 NY        PAS 2022-06-10   
1      1484697315  KEV4487                 NY        PAS 2022-06-13   
2      1484697625   H73NYD                 NJ        PAS 2022-06-19   
3      1484697674  GJC9296                 NY        PAS 2022-06-19   
4      1484697686   M51PUV                 NJ        PAS 2022-06-19   

   violation_code vehicle_body_type vehicle_make issuing_agency  street_code1  \
0              67               SDN        TOYOT              P         34330   
1              51              SUBN         JEEP              K         34310   
2              63               SDN         JEEP              N         30640   
3              63              SUBN        LEXUS              N         30640   
4              63               SDN        HYUND              N         30640   

   ...  feet_from_curb  violation_post_code  violation_description  \
0  ...               0                 None                   None   
1  ...               0                 None                   None   
2  ...               0                 None                   None   
3  ...               0                 None                   None   
4  ...               0                 None                   None   

   no_standing_or_stopping  hydrant_violation  double_parking_violation  \
0                     None               None                      None   
1                     None               None                      None   
2                     None               None                      None   
3                     None               None                      None   
4                     None               None                      None   

   violation_time_24hr day_of_week issue_month      Violation_address  
0      0 days 10:37:00      Friday        June  W 28TH ST, MANHATTHAN  
1      0 days 10:45:00      Monday        June    27TH DR, MANHATTHAN  
2      0 days 11:16:00      Sunday        June   SOUTH ST, MANHATTHAN  
3      0 days 10:52:00      Sunday        June   SOUTH ST, MANHATTHAN  
4      0 days 11:07:00      Sunday        June   SOUTH ST, MANHATTHAN  

[5 rows x 47 columns]

In [3]:
# Check the head of vio_22_time
query = """
SELECT 
    *
FROM 
    vio_22_time
LIMIT 10;
"""

df = execute_query(query)

# Display the DataFrame
df.head()


Plate State License_Type Summons_Number  Issue_Date Violation_Time  \
0  KLU1148    NY          PAS     4025715132  2022-08-09       07:04 AM   
1  JMT4956    NY          PAS     4025715181  2022-08-09       07:07 AM   
2  77157MH    NY          COM     1471746185  2022-03-24       03:07 PM   
3  EXH7265    NY          PAS     1471763055  2022-03-23       05:33 PM   
4  KVF7289    NY          PAS     4025716604  2022-08-09       08:27 AM   

                Violation Judgment_Entry_Date  Fine_Amount  Penalty_Amount  \
0      BUS LANE VIOLATION                            100.0             0.0   
1      BUS LANE VIOLATION                            150.0             0.0   
2  STORAGE-3HR COMMERCIAL                             65.0             0.0   
3          DOUBLE PARKING                            115.0             0.0   
4      BUS LANE VIOLATION                             50.0             0.0   

   ...  Amount_Due  Precinct    County                Issuing_Agency  \
0  ...         0.0         0  BROOKLYN  DEPARTMENT OF TRANSPORTATION   
1  ...         0.0         0     BRONX  DEPARTMENT OF TRANSPORTATION   
2  ...         0.0       107    QUEENS             POLICE DEPARTMENT   
3  ...         0.0       105    QUEENS             POLICE DEPARTMENT   
4  ...         0.0         0    QUEENS  DEPARTMENT OF TRANSPORTATION   

   Violation_Status                                      Summons_Image  \
0                    View Summons (http://nycserv.nyc.gov/NYCServWe...   
1                    View Summons (http://nycserv.nyc.gov/NYCServWe...   
2                    View Summons (http://nycserv.nyc.gov/NYCServWe...   
3                    View Summons (http://nycserv.nyc.gov/NYCServWe...   
4                    View Summons (http://nycserv.nyc.gov/NYCServWe...   

  Time_of_Violation Violation_Day Violation_Month violation_time_24hr  
0          07:04 AM       Tuesday          August     0 days 07:04:00  
1          07:07 AM       Tuesday          August     0 days 07:07:00  
2          03:07 AM      Thursday           March     0 days 15:07:00  
3          05:33 AM     Wednesday           March     0 days 17:33:00  
4          08:27 AM       Tuesday          August     0 days 08:27:00  

[5 rows x 23 columns]

# Data Cleaning and Preparation


In [4]:
# Verify that the date range of data in the vio_22_time table is correct.
query = """
SELECT 
    MIN(issue_date) AS earliest_date,
    MAX(issue_date) AS latest_date
FROM
    vio_22_time
WHERE
    issue_date is not null;
   
"""

df = execute_query(query)

# Display the DataFrame
df.head()

earliest_date latest_date
0    2022-01-01  2022-12-31

In [5]:
# Verify that the date range of data in the vio_22_time table is correct.
query = """
SELECT 
    MIN(issue_date) AS earliest_date,
    MAX(issue_date) AS latest_date
FROM
    parking_violations
WHERE
    issue_date is not null;
   
"""

df = execute_query(query)

# Display the DataFrame
df.head()

earliest_date latest_date
0    2022-01-02  2022-12-31

## Convert violation_time column to 24 hr format
   
The data in the `violation_time` column is entered as a string with "A" and "P" at the end to indicate 'AM' and 'PM'. I decided to convert the time to a 24-hour format for consistency and ease of analysis. 

First, a new column (`violation_time_24hr`) was created for the 24-hour time format. During the conversion process, some typographical errors were found in the entries. A regex was used to identify the malformed entries. The entries with typos were replaced with `NULL` values, and finally, the new column was updated with the 24-hour format. The code below was used to achieve this.    

In [6]:
query = "SELECT violation_time, violation_time_24hr FROM parking_violations LIMIT 10;"

df = execute_query(query)

# Display the DataFrame
df.head()

violation_time violation_time_24hr
0          1037A     0 days 10:37:00
1          1045A     0 days 10:45:00
2          1116A     0 days 11:16:00
3          1052A     0 days 10:52:00
4          1107A     0 days 11:07:00

## Extract the day of the week from the issue_date column

To determine how violations vary by day of the week, I decided to extract the day of the week from the `issue_date`. I used the `DAYNAME()` function to obtain the day name.

### Steps:
1. **Extracting Day of the Week:** Use the `DAYNAME()` function on the `issue_date` column to get the name of the day.
2. **Creating a New Column:** Store the extracted day names in a new column for further analysis.

In [7]:
# verify the day_of_week column creation

query = """
SELECT 
    day_of_week
FROM 
    parking_violations
LIMIT 10;
"""

df = execute_query(query)

# Display the DataFrame
df.head()

day_of_week
0      Friday
1      Monday
2      Sunday
3      Sunday
4      Sunday

In [8]:
# verify the issue_moth column creation

query = """
SELECT 
    issue_month
FROM 
    parking_violations
LIMIT 10;
"""

df = execute_query(query)

# Display the DataFrame
df.head()

issue_month
0        June
1        June
2        June
3        June
4        June

## Create a Column for the Violation Address

To explore the violation hotspots in NYC, I concatenated the entries in the `street_name` and `violation_county` columns of the `parking_violations` table to form a complete violation address. This information was stored in a new column called `violation_address`.

### Steps:
1. **Concatenation:** Combine the `street_name` and `violation_county` columns to create the `violation_address`.
2. **Creating a New Column:** Store the concatenated addresses in the `violation_address` column.    

In [9]:
# Verify that violation address column has been created.

query = """
SELECT 
    violation_address
FROM
    parking_violations;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()



violation_address
0  W 28TH ST, MANHATTHAN
1    27TH DR, MANHATTHAN
2   SOUTH ST, MANHATTHAN
3   SOUTH ST, MANHATTHAN
4   SOUTH ST, MANHATTHAN

## Convert Violation Addresses to Longitude and Latitude Coordinates

Once the violation address was obtained, I used the Geocode by Awesome Table extension in Google Drive to convert the entries in the `violation_address` column to longitude and latitude coordinates for use in Tableau.

### Steps:
1. **Geocoding:** Use the Geocode by Awesome Table extension to convert addresses to geographic coordinates.
2. **Storing Coordinates:** Save the resulting longitude and latitude coordinates for each address.

The geocoded data allows for more detailed spatial analysis in Tableau, enabling us to map the locations of parking violations and identify geographic patterns and hotspots.

By converting violation addresses to coordinates, we can create more insightful and interactive visualizations in Tableau, helping to better understand the spatial distribution of parking violations in NYC.


# Explore the data

## Descriptive Statistics

In [10]:
# Descriptive statistics for some of the numerical variables

query = """
SELECT 
    SUM(Fine_Amount) AS Fine_Amount_Sum,
    AVG(Fine_Amount) AS Fine_Amount_Mean,
    SUM(Penalty_Amount) AS Penalty_Amount_Sum,
    AVG(Penalty_Amount) AS Penalty_Amount_Mean,
    SUM(Interest_Amount) AS Interest_Amount_Sum,
    AVG(Interest_Amount) AS Interest_Amount_Mean,
    SUM(Reduction_Amount) AS Reduction_Amount_Sum,
    AVG(Reduction_Amount) AS Reduction_Amount_Mean,
    SUM(Payment_Amount) AS Payment_Amount_Sum,
    AVG(Payment_Amount) AS Payment_Amount_Mean,
    STDDEV(Fine_Amount) AS Fine_Amount_StdDev,
    STDDEV(Penalty_Amount) AS Penalty_Amount_StdDev,
    STDDEV(Interest_Amount) AS Interest_Amount_StdDev,
    STDDEV(Reduction_Amount) AS Reduction_Amount_StdDev,
    STDDEV(Payment_Amount) AS Payment_Amount_StdDev,
    MAX(Fine_Amount) AS Fine_Amount_Max,
    MIN(Fine_Amount) AS Fine_Amount_Min,
    MAX(Penalty_Amount) AS Penalty_Amount_Max,
    MIN(Penalty_Amount) AS Penalty_Amount_Min,
    MAX(Interest_Amount) AS Interest_Amount_Max,
    MIN(Interest_Amount) AS Interest_Amount_Min,
    MAX(Reduction_Amount) AS Reduction_Amount_Max,
    MIN(Reduction_Amount) AS Reduction_Amount_Min,
    MAX(Payment_Amount) AS Payment_Amount_Max,
    MIN(Reduction_Amount) AS Reduction_Amount_Min
FROM
    vio_22_time;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()


Fine_Amount_Sum  Fine_Amount_Mean  Penalty_Amount_Sum  Penalty_Amount_Mean  \
0     1.116999e+09         68.205012         188387125.0             11.50309   

   Interest_Amount_Sum  Interest_Amount_Mean  Reduction_Amount_Sum  \
0           5420776.88              0.330998           58387894.42   

   Reduction_Amount_Mean  Payment_Amount_Sum  Payment_Amount_Mean  ...  \
0               3.565218        8.232696e+08            50.269598  ...   

   Fine_Amount_Max  Fine_Amount_Min  Penalty_Amount_Max  Penalty_Amount_Min  \
0            515.0              0.0                60.0                 0.0   

   Interest_Amount_Max  Interest_Amount_Min  Reduction_Amount_Max  \
0                25.49                  0.0                900.43   

   Reduction_Amount_Min  Payment_Amount_Max  Reduction_Amount_Min  
0                   0.0              824.88                   0.0  

[1 rows x 25 columns]

The data above indicates that total fines amounted to about USD 1.12 billion, and the average fine issued was about 68.21 USD in 2022. 

## The top violations

In [11]:
# The Top 20 committed violations in 2022

query = """
SELECT 
    *
FROM
    top_violations;    
"""

df = execute_query(query)

# Display the DataFrame
df

Violation  Number_of_Violations
0   PHTO SCHOOL ZN SPEED VIOLATION               5747268
1       NO PARKING-STREET CLEANING               1837428
2   FAIL TO DSPLY MUNI METER RECPT               1070696
3      NO STANDING-DAY/TIME LIMITS                891144
4                     FIRE HYDRANT                743547
5       NO PARKING-DAY/TIME LIMITS                729790
6    INSP. STICKER-EXPIRED/MISSING                688653
7     FAILURE TO STOP AT RED LIGHT                619062
8               BUS LANE VIOLATION                602425
9     REG. STICKER-EXPIRED/MISSING                432275
10                  DOUBLE PARKING                353647
11            NO STANDING-BUS STOP                278716
12              EXPIRED MUNI METER                270038
13     FRONT OR BACK PLATE MISSING                261327
14     NO STANDING-COMM METER ZONE                250266
15  FAIL TO DISP. MUNI METER RECPT                229071
16  NO STANDING-EXC. TRUCK LOADING                223478
17  NO STANDING-EXC. AUTH. VEHICLE                132102
18    DOUBLE PARKING-MIDTOWN COMML                104551
19                       CROSSWALK                 97674

The results above show that the top three violations were:
1. **School Zone Speeding Violations:** 5.7 million
2. **No Parking Street Cleaning:** 1.8 million
3. **Failure to Display Municipal Meter Receipts:** 1.07 million

---

### Analysis Insights:
- **School Zone Speeding Violations:** These violations are the most frequent, indicating a high level of enforcement in school zones to ensure the safety of children.
- **No Parking Street Cleaning:** This suggests strict adherence to regulations for street cleanliness and maintenance.
- **Failure to Display Municipal Meter Receipts:** Highlights the importance of compliance with parking meter rules to avoid fines.

Understanding these top violations can help in formulating targeted interventions and policies to reduce their occurrence and improve overall traffic management in NYC.


## Violations by Registration State

In [12]:
# Violations by Registration State

query = """
SELECT 
    State, COUNT(State) AS Violations_By_Registration_State
FROM
    vio_22_time
WHERE
    State != '99' AND State != 'NF'
GROUP BY State
ORDER BY Violations_BY_Registration_State DESC;
"""

df = execute_query(query)

# Display the DataFrame
df.head()



State  Violations_By_Registration_State
0    NY                          12278101
1    NJ                           1495630
2    PA                            537441
3    FL                            362666
4    CT                            272919

Based on the vehicle registration state, the results indicate that the most violations were committed by vehicles from:
1. **New York (NY):** 12.3 million violations
2. **New Jersey (NJ):** 1.50 million violations
3. **Pennsylvania (PA):** 537,441 violations

---

### Analysis Insights:
- **New York (NY) Vehicles:** As expected, vehicles registered in NY state account for the majority of violations, likely due to the high number of NY-registered vehicles in NYC.
- **New Jersey (NJ) Vehicles:** The significant number of violations by NJ-registered vehicles suggests a high volume of commuters from NJ into NYC.
- **Pennsylvania (PA) Vehicles:** While lower than NY and NJ, the violations by PA-registered vehicles indicate that drivers from PA also frequently travel to and violate parking regulations in NYC.

These insights can help in understanding the demographic distribution of violators and tailoring enforcement strategies accordingly.


## Time series analysis of violations and fines

In [13]:
# How have violations and fines changed over time?
query = """
SELECT 
    Issue_Date,
    COUNT(*) AS Violations_by_Date,
    SUM(Fine_Amount) AS Total_Fine_Amount
FROM
    vio_22_time
GROUP BY Issue_Date
ORDER BY Violations_by_Date DESC;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

Issue_Date  Violations_by_Date  Total_Fine_Amount
0  2022-05-05               67108          4609675.0
1  2022-08-04               66332          4380930.0
2  2022-08-05               65088          4309100.0
3  2022-05-12               64824          4469020.0
4  2022-06-03               64707          4446150.0

In [14]:
# Are there specific times of the day when violations are more common?

query = """
SELECT 
    violation_time_24hr,
    COUNT(violation_time_24hr) AS time_of_day
FROM
    vio_22_time
GROUP BY violation_time_24hr
ORDER BY time_of_day DESC;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

violation_time_24hr  time_of_day
0     0 days 08:36:00        33207
1     0 days 08:39:00        31261
2     0 days 08:38:00        31088
3     0 days 08:40:00        30875
4     0 days 09:06:00        30595

Based on the data, most of the violations occur at around 8:36 AM. The number of violations starts rising from about 5:45 AM, peaks at 8:36 AM and then falls off after 9:00 AM. There is another peak at about 11:42 AM, after which the number of violations generally declines and peaks again at about 4:09 PM. This pattern is highlighted in the 'Total Parking and Traffic Violations Based on the Time of Day' visualization in the Tableau section of this project (place link here).

---

### Analysis Insights:
- **Early Morning Peak:** The highest number of violations occurs at 8:36 AM, suggesting a rush period likely associated with morning commutes.
- **Second Peak:** Another significant peak at 11:42 AM may be related to late morning activities or enforcement patterns.
Overall Trend:** After the second peak, the number of violations steadily declines and then peaks at about 4:09 PM, indicating the 4:00 PM rush hour. reasons for the general decline after 11:42 AM may be due to  reduced enforcement or lower traffic volumes in the afternoon.

These insights help in understanding the temporal distribution of violations, which can inform targeted enforcement strategies and better traffic management during peak hours.


In [15]:
## Violation By Day of the week

query = """
# Violations by day 
select
	Violation_Day, count(Violation_Day) as Violations_by_day
from
	vio_22_time
where 
	Violation_Day is not null
group by
	Violation_Day
order by
	Violations_by_day desc;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

Violation_Day  Violations_by_day
0      Thursday            2885322
1        Friday            2875594
2       Tuesday            2783433
3     Wednesday            2699529
4        Monday            2402448

The results show that most violations occurred on Friday (1.283 million), followed by Thursday (1.282 million). The least violations occurred on Sundays (694,597) and Saturdays (943,958). The 'NYC Parking and Traffic Violations by Day of the Week 2022' chart in Tableau highlights the results of this query.

---

### Analysis Insights:
- **Friday:** The highest number of violations occurs on Friday, indicating increased traffic and enforcement activity at the end of the workweek.
- **Thursday:** Close behind Friday, Thursday also shows a high number of violations, reflecting similar patterns of traffic and enforcement.
- **Weekend:** The lower number of violations on Sundays and Saturdays suggests reduced traffic volumes and enforcement activity during the weekend.

These insights can help in understanding the temporal distribution of parking and traffic violations, allowing for better planning and resource allocation for enforcement throughout the week.


In [16]:
# Violations by County

query = """
SELECT 
    County, COUNT(County) AS Violations_By_County
FROM
    vio_22_time
WHERE
    County != 'UNKNOWN' AND County != ''
GROUP BY County
ORDER BY Violations_By_County DESC;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

County  Violations_By_County
0       BROOKLYN               4533104
1     MANHATTHAN               4468515
2         QUEENS               4420998
3          BRONX               2320276
4  STATEN ISLAND                593968

Brooklyn and Manhattan were the counties with the most traffic violations. This insight is highlighted in the 'Parking and Traffic Violation Hotspots in NYC' chart in Tableau (place link here).

---

### Analysis Insights:
- **Brooklyn:** The high number of violations in Brooklyn indicates significant enforcement and/or high traffic volumes in this borough.
- **Manhattan:** Similarly, Manhattan's high violation count is likely due to its dense population and heavy traffic, leading to more frequent violations.

These insights can help in focusing enforcement efforts and addressing specific issues in these counties to improve compliance with parking and traffic regulations.

## Issuing agency analysis

In [17]:
# Which department made the most revenue for the city through parking fines?

query = """
SELECT 
    Issuing_Agency AS Agency, SUM(Fine_Amount) AS Fine_Amount
FROM
    vio_22_time
WHERE
    Issuing_Agency IS NOT NULL
GROUP BY Issuing_Agency
ORDER BY Fine_Amount DESC;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

Agency  Fine_Amount
0                       TRAFFIC  678349570.0
1  DEPARTMENT OF TRANSPORTATION  371736450.0
2      DEPARTMENT OF SANITATION   33943335.0
3             POLICE DEPARTMENT   28650933.0
4              PARKS DEPARTMENT    1558102.0

Among the issuing agencies, the Traffic Department issued the most fines in 2022 (678,349,570.00 USD), followed by the Department of Sanitation (371,736,450.00 USD), and the Police Department (28,650,933.00 USD).

---

### Analysis Insights:
- **Traffic Department:** The significant amount of fines issued by the Traffic Department highlights its primary role in enforcing parking regulations and traffic laws.
- **Department of Sanitation:** The high fines from the Department of Sanitation indicate a strong focus on maintaining cleanliness and adherence to sanitation-related parking rules.
- **Police Department:** Although lower in comparison, the fines issued by the Police Department reflect its enforcement of various traffic and parking violations.

These insights provide a clearer understanding of the contributions of different agencies to the overall enforcement efforts and financial impact of parking violations in NYC.


In [18]:
# Which issuing agency issued the most violations in 2022?
query = """
SELECT 
    Issuing_Agency, COUNT(Issuing_Agency) AS Agency
FROM
    vio_22_time
WHERE
    Issuing_Agency IS NOT NULL
GROUP BY Issuing_Agency
ORDER BY Agency DESC
LIMIT 20;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()



Issuing_Agency   Agency
0                       TRAFFIC  8530186
1  DEPARTMENT OF TRANSPORTATION  7016563
2      DEPARTMENT OF SANITATION   504934
3             POLICE DEPARTMENT   277312
4              PARKS DEPARTMENT    17225

In [19]:
# Which officers issued the most violation tickets in NYC in 2022

query = """
SELECT 
    Issuer_Code, COUNT(Issuer_Code) AS issued_violations
FROM
    parking_violations
WHERE
    Issuer_Code IS NOT NULL
        AND Issuer_Code != 0
GROUP BY Issuer_Code
ORDER BY issued_violations DESC
LIMIT 20;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

Issuer_Code  issued_violations
0       362237               9387
1       353177               9024
2       367329               8466
3       346199               8116
4       355542               7608

Based on the results, Officer number 362237 issued the most violation tickets (9378), followed by Officer 353117 and Officer 367329. If you are in the NYC area, be on the lookout for these officers—they are actively enforcing regulations!

---

### Analysis Insights:
- **Top Issuing Officers:** Officer 362237 issued the highest number of tickets, indicating a high level of activity and enforcement. Officers 353117 and 367329 also issued a significant number of tickets.
- **Enforcement Focus:** These officers' high ticket issuance suggests targeted enforcement efforts in specific areas or during certain times.

Understanding which officers are issuing the most tickets can provide insights into enforcement patterns and help in analyzing the effectiveness of traffic regulation enforcement in NYC.


## Vehicle Analysis
### How do vehicle brand color and body type relate to a number of violations? 
The entries for vehicle brand, color, and body type are inconsistent, so I first viewed the unique entries.

In [20]:
# unique entries 

query = """
SELECT DISTINCT
    vehicle_Make AS Make,
    Vehicle_Color AS Color,
    Vehicle_Body_Type AS Body_Type
FROM 
    parking_violations;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

Make Color Body_Type
0  TOYOT   BLK       SDN
1   JEEP  GRAY      SUBN
2   JEEP  GRAY       SDN
3  LEXUS  None      SUBN
4  HYUND  BLUE       SDN

In [21]:
# violation based on the brand of vehicles
query = """
SELECT 
    CASE 
        WHEN vehicle_make IN ('TOYOT', 'TOYOTA') THEN 'TOYOTA'
        WHEN vehicle_make IN ('JEEP') THEN 'JEEP'
        WHEN vehicle_make IN ('LEXUS') THEN 'LEXUS'
        WHEN vehicle_make IN ('HYUND', 'HYUNDAI') THEN 'HYUNDAI'
        WHEN vehicle_make IN ('NISSA', 'NISSAN') THEN 'NISSAN'
        WHEN vehicle_make IN ('HONDA') THEN 'HONDA'
        WHEN vehicle_make IN ('CHEVR', 'CHEVROLET') THEN 'CHEVROLET'
        WHEN vehicle_make IN ('FORD') THEN 'FORD'
        WHEN vehicle_make IN ('BMW') THEN 'BMW'
        WHEN vehicle_make IN ('DODGE', 'RAM') THEN 'DODGE'
        WHEN vehicle_make IN ('ACURA') THEN 'ACURA'
        WHEN vehicle_make IN ('SUBAR', 'SUBARU') THEN 'SUBARU'
        WHEN vehicle_make IN ('ME/BE') THEN 'MERCEDES BENZ'
        WHEN vehicle_make IN ('MITSU') THEN 'MITSUBISHI'
        WHEN vehicle_make IN ('INFIN') THEN 'INFINITI'
        WHEN vehicle_make IN ('VOLKS') THEN 'VOLKSWAGEN'
        WHEN vehicle_make IN ('MERCU') THEN 'MERCURY'
        WHEN vehicle_make IN ('ST', 'SMART') THEN 'SMART'
        ELSE vehicle_make
    END AS Vehicle_Brand, 
    COUNT(*) AS Violations_by_Vehicle_Make
FROM
    parking_violations
WHERE
    vehicle_make IS NOT NULL
GROUP BY 
    Vehicle_Brand
ORDER BY 
    Violations_by_Vehicle_Make DESC;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()


Vehicle_Brand  Violations_by_Vehicle_Make
0         HONDA                      915409
1        TOYOTA                      873959
2          FORD                      700503
3        NISSAN                      633665
4     CHEVROLET                      405079

In [22]:
# Which are the most popular streets for violations?
query = """
SELECT 
    violation_address,
    COUNT(violation_address) AS Violations_By_Street
FROM
    parking_violations
WHERE
    violation_address IS NOT NULL
GROUP BY violation_address
ORDER BY Violations_By_Street DESC
LIMIT 20;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

violation_address  Violations_By_Street
0    WB N CONDUIT AVE @ S, QUEENS                 71579
1            BROADWAY, MANHATTHAN                 65689
2  WB N. CONDUIT BLVD @, BROOKLYN                 59129
3             3RD AVE, MANHATTHAN                 49181
4     EB BRUCKNER BLVD @ W, BRONX                 38821

The results of this query were used to create the 'Parking and Traffic Violation Hotspots in NYC' visualization in Tableau. Here we see that some of the areas with the highest number of violations are WB N Conduit Ave @ S, Queens; Broadway, Manhattan; and WB N. Conduit Blvd @, Brooklyn.

---

### Analysis Insights:
- **WB N Conduit Ave @ S, Queens:** This location has a high number of violations, indicating a potential enforcement focus or frequent traffic issues.
- **Broadway, Manhattan:** The high violation count on Broadway reflects its busy nature and heavy traffic, leading to more frequent parking and traffic violations.
- **WB N. Conduit Blvd @, Brooklyn:** Similar to the Queens location, this area in Brooklyn also shows a high number of violations.

Identifying these hotspots can help in directing enforcement resources more effectively and addressing specific issues in these areas to improve traffic management and compliance.


# Dashboard Creation


I created a dashboard in Tableau to allow users to interact with the data. The dashboard presents the following KPIs: total violations, total fines issued, total fine payments, and the percentage of unpaid fines. The dashboard also includes charts highlighting how violations and fine payments vary over time, violations by vehicle registration state, the top issuing agencies based on fines, and the most frequently committed violations.

To create the dashboard, the data was aggregated based on the type of violations, fines, issuing agency, and registration state. The following queries were used to create four CSV files, which were then joined in Tableau to create the data source for the dashboard. Click here to view the dashboard (place link here).

---

### Key Dashboard Features:
- **Total Violations:** Displays the overall number of parking and traffic violations.
- **Total Fines Issued:** Shows the total amount of fines issued for violations.
- **Total Fine Payments:** Indicates the total payments made towards fines.
- **Percentage of Unpaid Fines:** Highlights the proportion of fines that remain unpaid.
- **Temporal Variation:** Charts showing how violations and fine payments change over time.
- **Violations by State:** Visualization of violations based on the state of vehicle registration.
- **Top Issuing Agencies:** Identifies the agencies issuing the most fines.
- **Top Committed Violations:** Lists the most common types of violations.

These features provide a comprehensive overview of parking and traffic violations in NYC, helping stakeholders understand patterns and trends to make informed decisions.


In [23]:
# Daily violations by type
query = """
SELECT 
    DATE(Issue_Date) AS date,
    Violation,
    COUNT(*) AS violation_count,
    SUM(Fine_Amount) AS total_fines
FROM
    vio_22_time
WHERE
    Issue_Date IS NOT NULL
        AND Violation IS NOT NULL
GROUP BY DATE(Issue_Date) , Violation;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

date                       Violation  violation_count  total_fines
0  2022-01-01             BEYOND MARKED SPACE                2         90.0
1  2022-01-01                       BIKE LANE                5        575.0
2  2022-01-01  COMML PLATES-UNALTERED VEHICLE                4        460.0
3  2022-01-01                       CROSSWALK               33       3795.0
4  2022-01-01                DETACHED TRAILER                5        285.0

In [24]:
# Daily violations by registration state

query = """
SELECT 
        DATE(Issue_Date) AS date,
        State,
        COUNT(*) AS state_violations,
        SUM(Fine_Amount) AS state_fines
    FROM
        vio_22_time
    WHERE
        Issue_Date IS NOT NULL
            AND State IS NOT NULL
    GROUP BY DATE(Issue_Date) , State;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()

date State  state_violations  state_fines
0  2022-01-01    99                31       2795.0
1  2022-01-01    AL                 6        560.0
2  2022-01-01    AR                 2        165.0
3  2022-01-01    AZ                 2        110.0
4  2022-01-01    CA                 9        690.0

In [25]:
# Daily violations by issuing agency
query = """
SELECT 
    DATE(Issue_Date) AS date,
    Issuing_Agency,
    COUNT(*) AS violations_count,
    SUM(Fine_Amount) AS total_fines
FROM
    vio_22_time
WHERE
    Issue_Date IS NOT NULL and Issuing_Agency is not null
GROUP BY DATE(Issue_Date) , Issuing_Agency;
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()


date                Issuing_Agency  violations_count  total_fines
0  2022-01-01      DEPARTMENT OF SANITATION                 9        695.0
1  2022-01-01  DEPARTMENT OF TRANSPORTATION               929      46450.0
2  2022-01-01               FIRE DEPARTMENT                 4        405.0
3  2022-01-01        OTHER/UNKNOWN AGENCIES                37       3290.0
4  2022-01-01              PARKS DEPARTMENT                 2        175.0

In [26]:
# Daily fines issued, Fine payment, and Unpaid fines

query = """
SELECT 
    DATE(Issue_Date) AS date,
    COUNT(*) AS total_violations,
    SUM(Fine_Amount) AS total_fines_issued,
    SUM(Amount_Due) AS total_unpaid_fines,
    SUM(Payment_Amount) AS total_paid_fines
FROM
    vio_22_time
WHERE
    issue_date IS NOT NULL
GROUP BY DATE(Issue_Date);
    
"""

df = execute_query(query)

# Display the DataFrame
df.head()


date  total_violations  total_fines_issued  total_unpaid_fines  \
0  2022-08-09             63007           4210895.0          1475152.49   
1  2022-03-24             55341           3853520.0          1159492.27   
2  2022-03-23             54186           3880635.0          1157614.66   
3  2022-09-22             35726           2285310.0          1024160.80   
4  2022-08-20             51226           3044215.0          1131938.21   

   total_paid_fines  
0        3316883.61  
1        3266704.60  
2        3228484.96  
3        1631584.13  
4        2432425.51

The results of the queries above were saved as four CSV files. These files were then loaded into Tableau Desktop software and joined on the common 'date' field. This data was used to create the 'New York City Parking and Traffic Violations Dashboard 2022' in Tableau.

---

### Process Overview:
1. **Query Results:** The output from the queries was saved as four separate CSV files.
2. **Loading Data:** The CSV files were imported into Tableau Desktop.
3. **Joining Data:** The files were joined on the common 'date' field to create a unified dataset.
4. **Dashboard Creation:** The combined data was used to build the 'New York City Parking and Traffic Violations Dashboard 2022' in Tableau.

This process allowed for an integrated and interactive dashboard that provides comprehensive insights into parking and traffic violations in NYC for the year 2022.


# Conclusion

This comprehensive project utilized SQL for data cleaning, preparation, and exploration, and Tableau for visualization and dashboard creation, to analyze NYC parking and traffic violation data for the year 2022. The analysis revealed the following insights:

* Approximately 16.3 million traffic and parking violations, amounting to $1.12 billion in fines, were issued in NYC in 2022.
* As expected, the majority of violations (12.3 million) were for vehicles registered in NY.
* Most violations were issued in the morning, peaking around 8:36 AM, with additional peaks at approximately 11:42 AM and 4:06 PM.
* The most violations occurred on Fridays, while the fewest violations were issued on Sundays and Saturdays.
* The Traffic Department was the agency that issued the most violation fines.
* The percentage of unpaid fines was approximately 38.42% for the year 2022.

The findings of this project provide valuable insights for both NYC city administrators and the general public in understanding the realities of parking and traffic violations in NYC. City administrators can use the results to identify areas for improvement, such as violation hotspots and the effectiveness of issuing agencies. The general public can use the findings to better understand the dynamics of parking and traffic violations in the city.
